File: .agents/skills/html-to-txt/convert.py
#!/usr/bin/env python3
"""
Convert a SingleFile HTML snapshot to clean .txt for e-readers.
Removes images, navigation, scripts, styles — outputs readable plain text.

Usage:  python3 convert.py file.html [--out FILE] [--width N]
"""

import argparse, os, re, sys
from html.parser import HTMLParser

BLOCK_TAGS = {'p','h1','h2','h3','h4','h5','h6','li','blockquote','pre'}
INLINE_TAGS = {'strong','b','em','i','u','s','a','span','sub','sup',
               'small','code','q','cite','abbr','dfn','kbd','mark'}
SKIP_TAGS  = {'script','style','nav','img','figure','svg','canvas',
              'noscript','iframe','video','audio','form','select',
              'button','input','textarea','object','embed','title'}
VOID_TAGS = {'img','input','br','hr','area','base','col','embed',
             'link','meta','param','source','track','wbr'}


class HtmlToText(HTMLParser):
    def __init__(self, width=80):
        super().__init__(convert_charrefs=True)
        self.width = width
        self._out = []
        self._buf = []
        self._skip = 0
        self._in_pre = False
        self._pre_buf = ''
        self._list_depth = 0

    # -- helpers -----------------------------------------------------------

    def _flush(self):
        t = ''.join(self._buf).strip()
        if t:
            self._out.append(t)
        self._buf = []

    def _writeln(self, line=''):
        self._out.append(line)

    def _wrap(self, text, indent=0):
        if self.width <= 0 or len(text) <= self.width:
            self._writeln(' ' * indent + text)
            return
        line, col = [], 0
        prefix = ' ' * indent
        for w in text.split():
            if col + len(w) + (1 if line else 0) > self.width:
                self._writeln(prefix + ' '.join(line))
                line, col = [w], len(w)
            else:
                line.append(w)
                col += len(w) + (1 if line else 0)
        if line:
            self._writeln(prefix + ' '.join(line))

    # -- parser ------------------------------------------------------------

    def handle_starttag(self, tag, attrs):
        tag = tag.lower()
        if self._skip:
            # Inside a skip region — only track non-void tags.
            if tag not in VOID_TAGS:
                self._skip += 1
            return
        if tag in SKIP_TAGS:
            self._skip = 0 if tag in VOID_TAGS else 1
            return

        # Inline tags should not trigger a block flush.
        if tag in INLINE_TAGS:
            return

        self._flush()

        if tag in ('h1','h2','h3','h4','h5','h6'):
            pass  # block tag, flush handled above
        elif tag == 'p':
            pass
        elif tag == 'blockquote':
            pass
        elif tag == 'li':
            pass
        elif tag in ('ul','ol'):
            self._list_depth += 1
        elif tag == 'br':
            self._writeln()
        elif tag == 'hr':
            self._writeln('─' * 40)
        elif tag == 'pre':
            self._in_pre = True
            self._pre_buf = ''

    def handle_endtag(self, tag):
        tag = tag.lower()
        if self._skip:
            self._skip -= 1
            return

        if tag == 'li':
            text = ''.join(self._buf).strip()
            self._buf = []
            if text:
                self._wrap(f'• {text}', indent=2 * self._list_depth - 2)
            return

        # Inline end tags: don't flush.
        if tag in INLINE_TAGS:
            return

        self._flush()

        if tag in ('p','h1','h2','h3','h4','h5','h6'):
            self._writeln()
        elif tag == 'blockquote':
            self._writeln()
        elif tag in ('ul','ol'):
            self._list_depth = max(0, self._list_depth - 1)
        elif tag == 'pre':
            self._in_pre = False
            t = re.sub(r'\n\s*\n', '\n', self._pre_buf.strip())
            if t:
                for line in t.split('\n'):
                    self._writeln('  ' + line)
                self._writeln()
            self._pre_buf = ''

    def handle_data(self, data):
        if self._skip:
            return
        if self._in_pre:
            self._pre_buf += data
            return
        self._buf.append(data)

    def handle_startendtag(self, tag, attrs):
        tag = tag.lower()
        if tag == 'br':
            self._writeln()
        elif tag == 'hr':
            self._writeln('─' * 40)
        elif tag == 'img':
            pass  # removed

    def get_text(self):
        self._flush()
        # Trim leading/trailing blank lines
        while self._out and self._out[-1] == '':
            self._out.pop()
        while self._out and self._out[0] == '':
            self._out.pop(0)
        return '\n'.join(self._out)


def convert(html_path, *, out_path=None, width=80):
    """Convert HTML file to .txt.  Returns the output path."""
    with open(html_path, 'r', encoding='utf-8', errors='replace') as f:
        raw = f.read()

    parser = HtmlToText(width=width)
    parser.feed(raw)
    text = parser.get_text()

    if out_path is None:
        out_path = os.path.splitext(html_path)[0] + '.txt'

    with open(out_path, 'w', encoding='utf-8') as f:
        f.write(text)
        f.write('\n')
    return out_path


def main():
    ap = argparse.ArgumentParser(
        description='Convert SingleFile HTML snapshot to clean .txt for e-readers.')
    ap.add_argument('html', help='Input HTML file')
    ap.add_argument('--out', help='Output path (default: same name, .txt)')
    ap.add_argument('--width', type=int, default=80,
                    help='Line wrap width (0 = disable, default: 80)')
    args = ap.parse_args()

    if not os.path.isfile(args.html):
        print(f'Error: file not found: {args.html}', file=sys.stderr)
        sys.exit(1)

    out = convert(args.html, out_path=args.out, width=args.width)
    print(f'Written: {out}')


if __name__ == '__main__':
    main()

File: .agents/skills/html-to-txt/convert.ts
#!/usr/bin/env npx tsx
/**
 * Convert a SingleFile HTML snapshot to clean .txt for e-readers.
 *
 * Uses Mozilla's Readability (Firefox Reader Mode) to extract the main
 * article content, then renders it as plain text — stripping all images,
 * navigation, sidebars, ads, and other boilerplate.
 *
 * Usage:  npx tsx convert.ts file.html [--out FILE] [--width N]
 */

import { JSDOM } from "jsdom";
import { Readability } from "@mozilla/readability";
import fs from "fs";
import path from "path";

// ---------------------------------------------------------------------------
//  HTML → Plain text (preserves paragraphs, headings, lists, blockquotes)
// ---------------------------------------------------------------------------

function htmlToText(html: string, width: number): string {
  const lines: string[] = [];
  const stack: string[] = [];
  let buf = "";

  function flush() {
    let t = buf.trim();
    // Clean up spaces before punctuation (artifacts from stripped inline tags)
    t = t.replace(/\s+([.,;:!?])/g, "$1");
    if (t) lines.push(t);
    buf = "";
  }

  // Tokenize via regex: opening tags, closing tags, text
  const re = /<(\/?)(\w+)[^>]*>|([^<]+)/gi;
  let m: RegExpExecArray | null;

  while ((m = re.exec(html)) !== null) {
    if (m[3] !== undefined) {
      // Text content
      const text = m[3]
        .replace(/&nbsp;/g, " ")
        .replace(/&amp;/g, "&")
        .replace(/&lt;/g, "<")
        .replace(/&gt;/g, ">")
        .replace(/&quot;/g, '"')
        .replace(/&#39;/g, "'")
        .replace(/&#(\d+);/g, (_s: string, n: string) =>
          String.fromCodePoint(parseInt(n, 10))
        )
        .replace(/\s+/g, " ")
        .trim();
      if (text) buf += (buf ? " " : "") + text;
      continue;
    }

    const isEnd = m[1] === "/";
    const tag = m[2].toLowerCase();

    if (isEnd) {
      const idx = stack.lastIndexOf(tag);
      if (idx !== -1) {
        // Pop the tag and process block-level formatting
        for (let i = stack.length - 1; i >= idx; i--) {
          const t = stack[i];
          if (t === "li") {
            flush();
            const text = buf.trim();
            buf = "";
            if (text) {
              const line = `• ${text}`;
              if (width > 0 && line.length > width) {
                wrapText(line, width, 2, lines);
              } else {
                lines.push(line);
              }
            }
          } else if (t === "p" || t.startsWith("h") || t === "blockquote") {
            flush();
            lines.push("");
          }
          // else inline tags: just pop, text stays in buf
        }
        stack.length = idx;
      }
    } else {
      stack.push(tag);
      if (tag === "br") {
        flush();
      } else if (tag === "hr") {
        flush();
        lines.push("─".repeat(40));
      }
    }
  }

  flush();

  // Trim trailing blank lines
  while (lines.length > 0 && lines[lines.length - 1] === "") lines.pop();
  // Trim leading blank lines
  while (lines.length > 0 && lines[0] === "") lines.shift();

  return lines.join("\n");
}

function wrapText(
  text: string,
  width: number,
  indent: number,
  out: string[]
): void {
  const prefix = " ".repeat(indent);
  let line: string[] = [];
  let col = 0;
  for (const word of text.split(" ")) {
    if (col + word.length + (line.length ? 1 : 0) > width - indent) {
      out.push(prefix + line.join(" "));
      line = [word];
      col = word.length;
    } else {
      line.push(word);
      col += word.length + (line.length > 1 ? 1 : 0);
    }
  }
  if (line.length) out.push(prefix + line.join(" "));
}

// ---------------------------------------------------------------------------
//  Main conversion
// ---------------------------------------------------------------------------

export function convert(
  htmlPath: string,
  options?: { outPath?: string; width?: number }
): string {
  const raw = fs.readFileSync(htmlPath, "utf-8");
  const width = options?.width ?? 80;
  const outPath =
    options?.outPath ?? htmlPath.replace(/\.html$/i, "") + ".txt";

  // Phase 1: extract readable content via Mozilla Readability
  // Use a strict threshold to minimise boilerplate bleeding through
  const dom = new JSDOM(raw, { url: "https://example.com" });
  const reader = new Readability(dom.window.document, {
    charThreshold: 150,
  });
  const article = reader.parse();

  let contentHtml: string;
  if (article?.content) {
    contentHtml = article.content;
  } else {
    // Fallback: use body content
    const body = dom.window.document.body;
    contentHtml = body ? body.innerHTML : raw;
  }

  // Strip remaining image/embed elements readability may leave
  contentHtml = contentHtml
    .replace(/<img[^>]*>/gi, "")
    .replace(/<figure[\s\S]*?<\/figure>/gi, "")
    .replace(/<svg[\s\S]*?<\/svg>/gi, "")
    .replace(/<canvas[\s\S]*?<\/canvas>/gi, "")
    .replace(/<video[\s\S]*?<\/video>/gi, "")
    .replace(/<audio[\s\S]*?<\/audio>/gi, "");

  const text = htmlToText(contentHtml, width);

  // Post-process: remove common boilerplate lines
  const boilerplate = [
    /^this site uses cookies/i,
    /^cookie policy/i,
    /^privacy policy/i,
    /^terms (of service|of use|conditions)/i,
    /^all rights reserved/i,
    /^©\s*\d{4}/i,
    /^subscribe now/i,
    /^sign up (for|to)/i,
    /^click to (read more|continue)/i,
    /^\x20{0,2}•\x20{0,2}$/,  // lone bullet point
  ];
  const cleanLines = text.split("\n").filter((line) => {
    const trimmed = line.trim();
    if (!trimmed) return true; // keep blank lines
    return !boilerplate.some((re) => re.test(trimmed));
  });

  const cleaned = cleanLines.join("\n");

  fs.writeFileSync(outPath, cleaned + "\n", "utf-8");
  return outPath;
}

// ---------------------------------------------------------------------------
//  CLI
// ---------------------------------------------------------------------------

function main(): void {
  const args = process.argv.slice(2);
  let htmlFile: string | undefined;
  let outFile: string | undefined;
  let width = 80;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--out" && i + 1 < args.length) {
      outFile = args[++i];
    } else if (args[i] === "--width" && i + 1 < args.length) {
      width = parseInt(args[++i], 10) || 80;
    } else if (!args[i].startsWith("--")) {
      htmlFile = args[i];
    }
  }

  if (!htmlFile) {
    console.error("Usage: npx tsx convert.ts file.html [--out FILE] [--width N]");
    process.exit(1);
  }

  if (!fs.existsSync(htmlFile)) {
    console.error(`Error: file not found: ${htmlFile}`);
    process.exit(1);
  }

  const out = convert(htmlFile, { outPath: outFile, width });
  console.log(`Written: ${out}`);
}

if (process.argv[1]?.endsWith("convert.ts")) {
  main();
}
